### PR TITLE
[TEP-0089] - Spire Package

### PR DESCRIPTION
**Pull request by pxp928**
Signed-off-by: pxp928 <parth.psu@gmail.com>

# Changes

Spire package separated out from https://github.com/tektoncd/pipeline/pull/4759 as requested. It includes the spire interface with a mocked spire for testing. As requested by @afrittoli.

/kind feature

# Submitter Checklist

As the author of this PR, please check off the items in this checklist:

- [ ] [Docs](https://github.com/tektoncd/community/blob/main/standards.md#docs) included if any changes are user facing
- [x] [Tests](https://github.com/tektoncd/community/blob/main/standards.md#tests) included if any functionality added or changed
- [x] Follows the [commit message standard](https://github.com/tektoncd/community/blob/main/standards.md#commits)
- [x] Meets the [Tekton contributor standards](https://github.com/tektoncd/community/blob/main/standards.md) (including
  functionality, content, code)
- [x] Release notes block below has been filled in
(if there are no user facing changes, use release note "NONE")

```release-note
Spire package that includes the controller and entrypointer API that is used in for [TEP-0089: Non-falsifiable provenance support](https://github.com/tektoncd/community/blob/main/teps/0089-nonfalsifiable-provenance-support.md). This is just the spire package. The phase 1 PR for the TEP is located here -> https://github.com/tektoncd/pipeline/pull/4759 
```

### CODE REVIEW

File: pkg/spire/controller.go
Comment (afrittoli) [inline pkg/spire/controller.go:98]: I would prefer to have two different factories here, one the tests and one for the actual usage, instead of embedding a `MockSpire` attribute in the client struct `spireControllerAPIClient`, it would be more consistent with all the other clients used in the controllers, like Tekton API clients, cloudevents client and cache client.

For instance, if you look at the https://github.com/tektoncd/pipeline/tree/main/pkg/reconciler/events/cache package, there are two clients defined there, the [regular one](https://github.com/tektoncd/pipeline/blob/main/pkg/reconciler/events/cache/cache.go) and the [mock one](https://github.com/tektoncd/pipeline/blob/main/pkg/reconciler/events/cache/cachefakeclient.go). 

The `knative/pkg` injection machinery makes sure that the correct client (regular or mock) is injected into the context, so that when you [get the client from the context](https://github.com/tektoncd/pipeline/blob/65d3bf3b2d1a828c4598ede08416aab899e2d779/pkg/reconciler/events/cloudevent/cloud_event_controller.go#L143) you get a different one, depending on whether it's unit tests or actual running code.

Comment (afrittoli) [reply pkg/spire/controller.go:98]: @mattmoor may want to have a say on this too 🙏

**Comment by afrittoli** [comment]
Thanks @pxp928 for moving this to a separate PR!

I'm reviewing the PR, but I find the module structure and test strategy not 100% clear:
In my understanding this is the role of each module is:

- *spire.go*: defines two interfaces, `ControllerAPIClient` and `EntrypointerAPIClient`. (NIT: maybe we call call this *clients*?)
- *controller.go*: implements some of `ControllerAPIClient`, including type, constructor and some of the required method, except for `VerifyTaskRunResults`
- *verify.go*: implements the rest of `ControllerAPIClient`, specifically `VerifyTaskRunResults` (NIT: maybe call this controller_verify?)
-  *entrypointer.go*: implements some of  `EntrypointerAPIClient`, including type, constructor and some of the required method, except for `Sign `
- *sign.go*: implements the rest of `EntrypointerAPIClient`, specifically `VerifyTaskRunResults`, specifically `VerifyTaskRunResults` (NIT: maybe call this controller_verify?)esults` (NIT: maybe call this entrypointer_sign?)
- *spire_mock.go*: defines a mock implementation of both interfaces. This bypasses most of the code in all the actual implementation, but it does invoke some of the actual functions, like `verifyManifest` and `getResultValue`
- *spire_mock_tests.go*: unit tests for *spire_mock.go*. These tests verify the logic in the mocks and some private functions in the *verify* and *sign* modules, but I find this a bit problematic:
  - it's very opaque what is tested and what not by the unit tests. This will make changing the code in future a rather convoluted task as changing private functions might break mock unit tests that depend on them. We have a coding guideline of testing only exported functions, and this is doing the opposite
  - there is no unit test coverage at all for a lot of the code defined in the actual implementations

The structure I would expect for this is something like a mock client, that implements only functions that already exists in the original [client](https://github.com/spiffe/go-spiffe/blob/main/v2/workloadapi/client.go), something like dial, close, get resource, list resources, create resource, delete resource. The `ControllerAPIClient` and `EntrypointerAPIClient` would only be responsible for implementing higher abstraction functions like  `VerifyTaskRunResults` and `Sign`. They would take a client as input, which would be the actual client at runtime, or the mock client at unit test time.

/cc @tektoncd/core-maintainers thoughts?

File: pkg/spire/controller.go
Comment (pxp928) [reply pkg/spire/controller.go:98]: Interesting. I did not know about this. I will update to use that.

File: pkg/spire/entrypointer.go
Comment (pxp928) [reply pkg/spire/entrypointer.go:35]: Changed to `setupClient` to better reflect the function.

File: pkg/spire/controller.go
Comment (pxp928) [reply pkg/spire/controller.go:98]: Hmm I see. I might need some help setting this up. I am not sure how this would work in this context.

File: pkg/spire/verify.go
Comment (wlynch) [inline pkg/spire/verify.go:189]: This doesn't make a functional difference, but should we use this for better type hinting from the func args? (also for verifyResult)

```suggestion
func verifySignature(pub crypto.PublicKey, signature string, value string) error {
```

**Review by wlynch — CHANGES_REQUESTED** [review]

File: pkg/spire/config/config.go
Comment (vdemeester) [inline pkg/spire/config/config.go:60]: I don't think the error is correct here (no image involevd)

**Review by vdemeester — COMMENTED** [review]
Agreeing with @wlynch and @afrittoli, in term of code organization especially for test.

File: pkg/spire/controller.go
Comment (pxp928) [reply pkg/spire/controller.go:28]: No reason. Accidental. Fixed.

File: pkg/spire/entrypointer.go
Comment (pxp928) [reply pkg/spire/entrypointer.go:62]: For this we are checking fetching the SVID every few seconds. The change suggested with break the logic. I changed the variable name to `backoff` as it might be the causing confusion.

File: pkg/spire/sign.go
Comment (pxp928) [reply pkg/spire/sign.go:43]: Agreed. Fixed.

File: pkg/spire/spire.go
Comment (pxp928) [reply pkg/spire/spire.go:42]: This TTL is for the how long the entries that are registered into the Spire Sever are valid. Currently we are also using the default timeout setup by the config when initializing it and it will be deleted when the `tr.isDone()`. 

```
// TTL for the entry is in seconds
ttl := config.FromContextOrDefaults(ctx).Defaults.DefaultTimeoutMinutes * 60
```

File: pkg/spire/verify.go
Comment (pxp928) [reply pkg/spire/verify.go:116]: A trust bundle is a collection of one or more certificate authority (CA) root certificates that the workload should consider trustworthy. Added the `len(x509Bundle)` check.

File: pkg/spire/verify.go
Comment (pxp928) [reply pkg/spire/verify.go:189]: Agreed. Changed.

File: pkg/spire/spire_mock.go
Comment (lumjjb) [reply pkg/spire/spire_mock.go:32]: Hi! Sorry about being late to reply. COVID happened :|

Wanted to share a bit more about the thought process around the mocking,

---
### For unit testing

I think the client mocking will work perfectly with unit tests.

----

### For integration

We explored several different approaches when trying to mock SPIRE, a lot of the difficulty in doing so was that test that use mocked SPIRE resulted in two separate processes interacting with the shared state for integration tests. Mocking the workload API for a single memory process is ok, but when two processes (in this case the entrypointer and the controller) need to be able to retrieve compatible crypto material, is where things get a bit messy since we need to maintain state across processes and isolated containers. We would have to mock a daemon server, which pretty much re-implements a bundled version of SPIRE.

The workaround was performing mocking at a slightly higher layer, which takes away the need to have shared state between 2 processes with a fixed seed which would result in the need of an external daemon.

I think there's a desire to have the mocked SPIRE daemon with compatible fake workload client, but it is something that will be at least a couple months work of a dedicated developer.

Another possibility here could also be to create determinism within the ecdsa/rsa crypto library, it's not something I've explored too much.

---

What are some thoughts here? Maybe client mocking for unit tests but keep the higher level mocking for now for the integration/e2e tests - until the SPIRE community makes more progress on a mock server?

File: pkg/spire/entrypointer.go
Comment (wlynch) [reply pkg/spire/entrypointer.go]: My bad - you're 100% right! Ignore this.

File: pkg/spire/spire.go
Comment (wlynch) [reply pkg/spire/spire.go:50]: My thinking in making it a duration is that we just make the above line

```
ttl := config.FromContextOrDefaults(ctx).Defaults.DefaultTimeoutMinutes * time.Minute
```

instead of converting to seconds. wdyt?

File: pkg/spire/spire_mock.go
Comment (wlynch) [reply pkg/spire/spire_mock.go:32]: No worries! Hope you're feeling better! 🙏 

That sounds good to me! My comment was more about which clients we are mocking for unit tests so that we cover more codepaths for the real implementation (in particular, I'm worried we might be missing stuff in sign.go / controller.go because spire_mock.go is using its own impl). I'd be okay with an additional higher level mock for e2e/integration tests if needed.

We may also want e2e tests that set up real spire, but I'm fine with making that out of scope for this PR.

File: pkg/spire/verify.go
Comment (wlynch) [reply pkg/spire/verify.go]: I could have phrased that better - I meant should we looping over `x509Bundle` (as opposed to just the authorities in first one) i.e.

```go
trustPool := x509.NewCertPool()
for _, bundle := range x509Bundle {
	for _, c := range bundle.X509Authorities() {
		trustPool.AddCert(c)
	}
	return trustPool, nil
}
```

File: pkg/spire/spire_mock.go
Comment (pxp928) [reply pkg/spire/spire_mock.go:32]: The e2e/integration tests already do set up the real spire. We have all this completed in the following two PRs (Phase 1 and Phase 2): https://github.com/tektoncd/pipeline/pull/4759 and https://github.com/tektoncd/pipeline/pull/4828

We are only using the mocked spire for unit testing with the various other packages (for example the pod and status).

Comment (lumjjb) [reply pkg/spire/spire_mock.go:32]: Oh sorry on that last "e2e" left that in there, meant to be just integration*. 

Ok so @wlynch so sounds like the plan is that we will add in the fakeworkload client for unit testing, for additional coverage and leave the alternative mocking (and use the injection mechanism instead) for integration tests? Does that sound about right?

Comment (wlynch) [reply pkg/spire/spire_mock.go:32]: Sounds great to me - thanks!

Comment (afrittoli) [reply pkg/spire/spire_mock.go:32]: I definitely would be happy to have mocks on the workload API for unit test.
For E2E tests I would tend to say we should use the full server instead of a mock.

File: pkg/spire/spire.go
Comment (lumjjb) [reply pkg/spire/spire.go:50]: it does end up being translated back to `int` when talking to the SPIRE API, but am okay with either.

File: pkg/spire/controller.go
Comment (lumjjb) [reply pkg/spire/controller.go:96]: seems do-able, does this look about right? https://github.com/pxp928/pipeline/pull/26/files

Comment (afrittoli) [reply pkg/spire/controller.go:96]: Thanks @lumjjb - that's definitely in the right direction! 

The only question I have is whether we could get the config from the context, so that when the client are setup the config is added directly, and we don't need to add it later on. Ideally we should add the client config once, unless we need to support reloading configuration changes at run time. Wdyt?

File: pkg/spire/verify.go
Comment (pxp928) [reply pkg/spire/verify.go]: Yup. Added.

File: pkg/spire/spire.go
Comment (pxp928) [reply pkg/spire/spire.go:50]: Changed to `time.Duration` but like Brandon said it will have to translate back to `int` for the SPIRE API.

File: pkg/spire/spire.go
Comment (pxp928) [reply pkg/spire/spire.go:55]: same question as above.

File: pkg/spire/controller.go
Comment (lumjjb) [reply pkg/spire/controller.go:96]: Thanks @afrittoli ! It's in a weird spot with the [initialization of the TaskRun reconciler](https://github.com/tektoncd/pipeline/blob/main/pkg/reconciler/taskrun/controller.go#L47), since some of the options come in as CLI arguments. Do you have any recommendations to handle this?

**Comment by pxp928** [comment]
@vdemeester @afrittoli @wlynch Should be ready for another review. @lumjjb added the injection piece and I have added the `fakeworkloadapi` unit tests to test the spire "sign" and "verify" functionality. I have also added some missing pieces to the spire package that are needed for phase 2 (signed TaskRun). This will allow the other two follow on PRs to utilize all the functionality from the spire package for TaskRun results and status signing and verification.

Once the review is complete. I will squash all the commits into one before it gets merged.

**Comment by pxp928** [comment]
[@vdemeester](https://github.com/vdemeester) [@afrittoli](https://github.com/afrittoli) [@wlynch](https://github.com/wlynch) Should be ready for another review. [@lumjjb](https://github.com/lumjjb) added the injection piece and I have added the fakeworkloadapi unit tests to test the spire "sign" and "verify" functionality. I have also added some missing pieces to the spire package that are needed for phase 2 (signed TaskRun). This will allow the other two follow on PRs to utilize all the functionality from the spire package for TaskRun results and status signing and verification.

Once the review is complete. I will squash all the commits into one before it gets merged.

File: .golangci.yml
Comment (pxp928) [reply .golangci.yml:47]: The spire test package had a bunch of linting errors but as it was copied over from spire, I was not sure if I made sense to fix it all.

**Comment by pxp928** [comment]
@afrittoli Also seems like we need your help to figure out how to initialize the context for client injection during initialization of the entrypoint, controller (cmd/entrypoint/main.go) etc.

File: pkg/spire/spire.go
Comment (pxp928) [reply pkg/spire/spire.go:58]: Added the links

File: pkg/spire/spire.go
Comment (pxp928) [reply pkg/spire/spire.go:70]: Added the links

File: pkg/spire/test/fakeworkloadapi/workload_api_windows.go
Comment (wlynch) [inline pkg/spire/test/fakeworkloadapi/workload_api_windows.go:2]: Can we add a short blurb on why a windows version is required? e.g. do we need to use the winio pipe, or can we simplify this with basic tcp on localhost?

File: .golangci.yml
Comment (wlynch) [reply .golangci.yml:47]: Is this a verbatim copy, or are we making modifications?

(if we're making modifications, I'd prefer to keep linting on for our changes if possible)

we should also add instructions for how this was imported for if/when we need to update.

File: pkg/spire/sign.go
Comment (wlynch) [inline pkg/spire/sign.go:138]: This could possibly also be len == 0? (probably worth having a sanity check here just in case).

File: pkg/spire/sign.go
Comment (wlynch) [reply pkg/spire/sign.go:46]: nit: I'd move the len check to here - it's not immediately obvious that it's being checked in getWorkloadSVID.

File: pkg/spire/spire.go
Comment (wlynch) [inline pkg/spire/spire.go:45]:
```suggestion
	VerifiedAnnotation = "tekton.dev/spire-verified"
```

A general rule of thumb I like to follow is avoiding double negatives. It helps avoid ambiguous cases like does the absence of this annotation mean that it was successfully verified?

Also recommend using `spire-verified` so this doesn't end up conflicting with other verification checks.

**Review by wlynch — COMMENTED** [review]
Looking good! Definitely digging this structure much more. 😎

File: pkg/spire/test/fakeworkloadapi/workload_api_windows.go
Comment (pxp928) [reply pkg/spire/test/fakeworkloadapi/workload_api_windows.go:2]: This is the remanent of the copy of the fakeworkloadapi. I will remove it.

**Comment by pxp928** [comment]
> Is this a verbatim copy, or are we making modifications?

Yes the `pkg/spire/test` folder is a verbatim copy from the upstream project. No modification were made.

File: pkg/spire/spire_mock.go
Comment (lumjjb) [reply pkg/spire/spire_mock.go:48]: Unfortunately the `MockClient` uses a lot of the `pkg/spire` internal helper functions. We could add a comment that this struct should not be used outside `tektoncd` testing. As another measure, we could also add a build tag for integration `// +build integration`. wdyt?

File: .golangci.yml
Comment (pritidesai) [reply .golangci.yml:47]: If the test client needs to be part of the pipeline code base, we will need license headers in each file under `pkg/spire/test`, fix linting errors, and some kind of documentation on why ...

File: pkg/reconciler/testing/logger.go
Comment (pritidesai) [inline pkg/reconciler/testing/logger.go:34]: NIT: extra `the` in `sets up the the Context`

File: pkg/reconciler/testing/logger.go
Comment (pritidesai) [inline pkg/reconciler/testing/logger.go:65]: NIT: extra `the` in `sets up the the Context`

File: pkg/reconciler/testing/logger.go
Comment (pritidesai) [inline pkg/reconciler/testing/logger.go:67]: just wondering how `setupDefaultContextWithLabelKey` is different from `setupFakeContextWithLabelKey`? it looks same 🤔 

https://github.com/tektoncd/pipeline/blob/00b14d78ad73c055451f2a714387cd03f6222b89/pkg/reconciler/testing/logger.go#L47-L53

File: pkg/reconciler/testing/logger.go
Comment (pritidesai) [inline pkg/reconciler/testing/logger.go:35]: same as `SetupFakeContext`:

https://github.com/tektoncd/pipeline/blob/00b14d78ad73c055451f2a714387cd03f6222b89/pkg/reconciler/testing/logger.go#L25-L32

File: pkg/spire/spire_test.go
Comment (pxp928) [reply pkg/spire/spire_test.go:68]: Fixed.

File: pkg/spire/sign.go
Comment (pxp928) [reply pkg/spire/sign.go:138]: Added check:
```
	if len(xsvid.Certificates) == 0 {
		return errors.New("returned controller svid does not have certificates")
	}
```

File: pkg/reconciler/testing/logger.go
Comment (pxp928) [reply pkg/reconciler/testing/logger.go:34]: fixed

File: pkg/reconciler/testing/logger.go
Comment (pxp928) [reply pkg/reconciler/testing/logger.go:65]: fixed.

File: pkg/reconciler/testing/logger.go
Comment (pxp928) [reply pkg/reconciler/testing/logger.go:67]: Clean up how the functions are called but the difference is the  `injection.Fake` to `injection.Default`. This is needed so that the fakeworkloadAPI can be called for unit testing the spire api within `spire_test.go`.

File: pkg/reconciler/testing/logger.go
Comment (pxp928) [reply pkg/reconciler/testing/logger.go:35]: Removed `cloudEventClientBehaviour` as not needed but this is to call the above `setupDefaultContextWithLabelKey` used to test spire with the fakeworkloadAPI.

File: pkg/spire/spire.go
Comment (pxp928) [reply pkg/spire/spire.go]: Agreed. Changed to suggested.

File: pkg/spire/config/config.go
Comment (pritidesai) [reply pkg/spire/config/config.go:60]: I think it will be great to make this error more descriptive. This error will be met when `SpireConfig` is missing all four required flags which in turn means no image initialized 🤔

File: .golangci.yml
Comment (pxp928) [reply .golangci.yml:47]: It is verbatim copy so do we need liniting?

File: pkg/spire/config/config.go
Comment (pxp928) [reply pkg/spire/config/config.go]: changed to "found unset spire configuration flags"

File: .golangci.yml
Comment (pxp928) [reply .golangci.yml:47]: Created an issue #5281 to track this.

**Comment by lumjjb** [comment]
@wlynch @pritidesai changes made PTAL!

**Review by wlynch — COMMENTED** [review]
I think this is pretty much there! /lgtm

File: pkg/spire/test/fakeworkloadapi/workload_api_posix.go
Comment (pxp928) [reply pkg/spire/test/fakeworkloadapi/workload_api_posix.go]: removed.

**Comment by lumjjb** [comment]
Hi @pritidesai we've made your requested! Can you take a look at the PR again - would like to get this across the line! Thanks!

**Review by wlynch — APPROVED** [review]
/lgtm

@pritidesai can you take another look at this for approval? 👀

**Review by vdemeester — APPROVED** [review]
/hold
@pritidesai @afrittoli one last review before removing the hold 🙏🏼

**Comment by wlynch** [comment]
Since this PR already has an LGTM from me and has been sitting without additional feedback for weeks, I'm going to go ahead and remove the hold to let this submit.

Nothing in this PR is a one-way door decision and is not wired up to the controller yet, so there's not much risk in submitting this even if there is additional feedback. Feel free to continue to leave feedback on this PR - we can follow up with additional PRs if needed!

/remove-hold

**Comment by afrittoli** [comment]
Great to see this moving forward! I'm sorry I wasn't able to re-review before, but I'm glad to see progress.

**Pull request merged**
